Protect card: Add the Scan/Threats status column (with tooltip)

elliottprogrammer opened the pull request:
This PR populates the Scan/Threats section on the My Jetpack Protect card with the current scan status from Protect.

<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->
See Design P2: p1HpG7-rFA-p2 (Figma link is within the post)

Screenshots are shown below in the testing instructions.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Move the info tooltip into it's own component.
* Create a component for the Scan/Threats section, rendering the current Protect scan status, per Figma design

**Note:** Adding a "loading/currently scanning" state in the My Jetpack Protect product card is not included in this PR and will be handled in a future PR.  We need to add the Scan API js-package first.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

Project Thread: pbNhbs-aP6-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

You can reference the link to the Figma design in the design P2 post: p1HpG7-rFA-p
(tampermonkey does not shorthand the direct Figma link for some reason, so you'll need to get it from the design post).

- Checkout this branch via the Jetpack Beta plugin or your local dev environment (In this case I would suggest Jetpack Beta so you can more easily start with a fresh site)
    - In the Jetpack Beta plugin, run this branch on the Jetpack plugin **and** the Protect Plugin
- Open your JN site to get the SSH credentials, then SSH into the JN site.
- OR: Edit the plugin file in the editor by going to: `/wp-admin/plugin-editor.php`
- Edit the Akismet plugin file to trigger a threat:
    - Edit the `akismet.php` file: ` vim htdocs/wp-content-plugins/akismet/akismet.php`
    - In the top comment section, change the "`Version:`" to: `3.1.4`
    - Also on line 40, change it to: `define( 'AKISMET_VERSION', '3.1.4' );`
- Open the My Jetpack page.
- Look at the Protect product card. The "Scan" status column should show "Off", since the site is not connected: (See screenshot):
![Screen Shot on 2024-07-05 at 11-22-27](https://github.com/Automattic/jetpack/assets/11078128/94e57120-2be4-432b-950d-72edbcbb49e7)
- Click "Protect" in the Jetpack menu to go to the Protect page.  Select the Free plan and wait for the scan to complete.
**Note:** Adding a "loading/currently scanning" state in the My Jetpack Protect product card is not included in this PR and will be handled in a future PR.
- Once the scan has completed, go to My Jetpack (or refresh the My Jetpack page). 
- You should see 1 Threat in the Protect product card and it should match the Figma design.
![Screen Shot on 2024-07-10 at 23-36-40](https://github.com/Automattic/jetpack/assets/11078128/6bca4713-1ed9-4ed2-baa4-befd88080cea)
- Now edit the Akismet plugin file again (`akismet.php`):
- Add the following line anywhere inside the file, and then save:
        - `$eicar = "X5O!P%@AP[4\PZX54(P^)7CC)7}$EICAR-CRITICAL-ANTIVIRUS-TEST-FILE!$H+H*";`
- Go to the Protect page.  Click "Upgrade Jetpack Protect now" and purchase with credits.
- After upgrade purchased, Protect should be scanning the site again.  Wait for the scan to complete.
Again note: adding a "loading/currently scanning" state in the My Jetpack Protect product card is not included in this PR and will be handled in a future PR.
- Once scan is completed, go to My Jetpack and view the Protect card.  It should show 1 threat (1 Critical threat), with a tooltip. (See screenshot):
![Screen Shot on 2024-07-04 at 15-03-56](https://github.com/Automattic/jetpack/assets/11078128/419ae970-60cc-4484-98cf-98f1a7df8561)
- Go to the protect page, and click to "Auto fix" the threat.  Wait for the threat fix to complete successfully.
- Once the threat is fixed, go to the My Jetpack page. View the protect card. It should show "Secure". (See screenshot):
![Screen Shot on 2024-07-05 at 12-37-54](https://github.com/Automattic/jetpack/assets/11078128/80c5a35a-f045-4a19-9a02-6183634f254b)
- Go to wordpress.com Store Admin and remove the Scan subscription you purchased for the site.
- Go back to the My Jetpack page and view the Protect card. It should say "Partial" with tooltip (See screenshot):
![Screen Shot on 2024-07-05 at 12-45-42](https://github.com/Automattic/jetpack/assets/11078128/7b0a390c-a0d5-4371-96d8-638529b9700b)
- Go to the Plugins page, and deactivate Protect.
- Go back to My Jetpack and the Protect card should Show "Off" (See screenshot):
![Screen Shot on 2024-07-05 at 12-50-39](https://github.com/Automattic/jetpack/assets/11078128/218ca06c-c8ed-43a1-bb60-3a730790f6f7)

elliottprogrammer committed:
Protect card: Add the Last scan time (with tooltip) based on plugin state.

elliottprogrammer committed:
Fix "Lern more" typo in comment, per feedback.

elliottprogrammer committed:
Add defaults to destructured var, per feedback.

elliottprogrammer committed:
Use React "FC" type, per feedback.

elliottprogrammer committed:
Fix snake case naming & state naming, per feedback.

elliottprogrammer committed:
CSS style fixes/updates, per feedback.

elliottprogrammer committed:
Show dynamic data in tooltips, per feedback; And cleanup/refactor.

elliottprogrammer committed:
Close tooltip/popover when outside focus, per feedback.

elliottprogrammer committed:
Fix tooltip close fires twice when tooltip focus switches to button clicked.

elliottprogrammer committed:
Fix responsiveness, per feedback.

elliottprogrammer committed:
Only show "Last scan:" if value available, per feedback.

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-protect-card-last-scan-time

elliottprogrammer committed:
Add cursor: pointer to tooltip button, per feedback.

elliottprogrammer committed:
Use --font-body-extra-small & fix comment typo, per feedback.

elliottprogrammer committed:
Only show "Plugins & Themes" when scan data availavle, per feedback.

elliottprogrammer committed:
Fixup project versions.

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-protect-card-last-scan-time

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-protect-card-last-scan-time

elliottprogrammer committed:
Allow for singular/plural in translations.

elliottprogrammer committed:
Move Popover into its own InfoTooltip component.

elliottprogrammer committed:
Protect card: Add the Scan/Threats status column (with tooltip)

elliottprogrammer committed:
Critical threats, tooltip tracking, & additional styles.

elliottprogrammer committed:
Fix core.threats not iterable error.

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-protect-card-last-scan-time

elliottprogrammer committed:
Merge branch 'add/mj-protect-card-last-scan-time' into add/my-protect-card-scan-threats

elliottprogrammer committed:
changelog.

elliottprogrammer committed:
Scan status "Off" when site not connected.

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/protect-value-section.tsx:
```diff
@@ -39,11 +38,21 @@ const ProtectValueSection = () => {
+		return (
+			sprintf(
+				/* translators: %d is the number of plugins installed on the site. */
+				_n( '%d plugin', '%d plugins', pluginsCount, 'jetpack-my-jetpack' ).replace( ' ', '\xa0' ), // `\xa0` is a non-breaking space.
```
If what we care about is to always have one line text, we can consider setting it in CSS (`white-space: nowrap;`) instead of HTML.

elliottprogrammer replied:
Thanks @robertsreberski but the goal here isn't to keep everything on one line, I'm just preventing the line from breaking between then number/digit and the following text, for example I don't want the text "5 plugins & 3 themes" to break between "5" and "plugins" or between "3" and "themes".  That way if it has to break into two lines, it will break like this:
5 plugins
& 3 themes

but not:
5 plugins & 3
themes

Anyway, that was the intention here and the reason for the specifically placed non-breaking spaces.
Although realizing now I could probably remove the non-breaking space between "5" and "plugins" (and keep the nbsp between "3" and "themes") because I don't think the card will ever squeeze small enough that it would wan't to break after the very first number, like between "5" and "plugins" in the text "5 plugins & 3 themes".

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/protect-value-section.tsx:
```diff
@@ -65,59 +74,30 @@ const ValueSection: FC< {
+						} }
+					>
+						<>
+							<h3 className="value-section__tooltip-heading">{ pluginsThemesTooltip.title }</h3>
+							<p className="value-section__tooltip-content">{ pluginsThemesTooltip.text }</p>
```
Does it make sense to also standardize the content and move the components to <InfoTooltip />? Or we rather keep the flexibility (will we need it?)

elliottprogrammer replied:
Personally, I don't see any specific beneficial reasoning for creating a separate component for the contents of the tooltip, especially since it's only a title and a bit of text.  Plus I do believe it's possible we might need the flexibility. I could easily see us wanting to add a link or a CTA button or something like that in there in the future.

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/style.scss:
```diff
@@ -57,4 +45,53 @@
+
+.scan-threats {
+    &__threat-count {
+        font-size: 32px;
```
In this case, I think it actually makes sense to still use the variable, just subtract the pixels (so to keep it "tied" to line-height). E.g. `calc( var(--font-title-large) - 4px )`

elliottprogrammer replied:
Agreed, nice suggestion, thanks!! 😃

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/style.scss:
```diff
@@ -57,4 +45,53 @@
+
+    &__status-icon {
+        display: block;
+        width: 18.6px;
```
If the value is so specific, we can leave some comment why we've decided to go with it (I understand it was in design?)

elliottprogrammer replied:
Thanks Robert, actually I'm realizing I don't even need to specifically set the icon size dimensions here. I can just let the icon image display at it's natural size. 👍. I'll remove the width and height definitions here.

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/use-protect-tooltip-copy.ts:
```diff
@@ -35,28 +35,49 @@ export function useProtectTooltipCopy( {
-			text: sprintf(
-				/* translators: %1$d is the number of plugins and %2$d is the number of themes installed on the site. */
+			text:
+				sprintf(
+					/* translators: %d is the number of plugins installed on the site. */
+					_n(
+						'Your site has %d plugin',
+						'Your site has %d plugins',
+						pluginsCount,
+						'jetpack-my-jetpack'
+					),
+					pluginsCount
+				) +
+				' ' +
+				sprintf(
+					/* translators: %d is the number of themes installed on the site. */
+					_n( 'and %d theme', 'and %d themes', themesCount, 'jetpack-my-jetpack' ),
+					themesCount
+				) +
+				' ' +
 				__(
-					'Your site has %1$d plugins and %2$d themes lacking security measures. Improve your site’s safety by adding protection at no cost.',
+					'lacking security measures. Improve your site’s safety by adding protection at no cost.',
 					'jetpack-my-jetpack'
 				),
```
What is the idea behind breaking it down to couple of strings? I'm open to it just keeping it as one still might have sense, like translators will have more context of the text they are translating. And we would avoid the situation text (especially since is one sentence) is partially translated.

elliottprogrammer replied:
@robertsreberski, Hmm yeah, I guess I just don't know how to use the `_n()` function for 2 different singular/plural items and have them all in one translated string.. 🤔🤷‍♂️. I'm not quite sure how I would do that?..
Care to give it a try, and show me how to do it? Or give me some pointers on how I would do that?

elliottprogrammer replied:
Nevermind, with @CodeyGuyDylan 's help, he showed me how to combine singular/plural translations by nesting the sprintf's.  Ok, should be all fixed now.

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/scan-threats.tsx:
```diff
@@ -0,0 +1,213 @@
+						/>
+					</div>
+					<div className="value-section__status-text">
+						{ __( 'Partial', 'jetpack-my-jetpack' ) }
+					</div>
+					<InfoTooltip
+						tracksEventName={ 'protect_card_tooltip_open' }
+						tracksEventProps={ {
+							location: 'scan',
+							status: 'partial',
+							hasPaidPlan: false,
+							threats: 0,
+						} }
+					>
+						<>
+							<h3 className="value-section__tooltip-heading">{ scanThreatsTooltip.title }</h3>
+							<p className="value-section__tooltip-content">{ scanThreatsTooltip.text }</p>
+						</>
+					</InfoTooltip>
+				</div>
+			</>
+		);
+	}
+
+	return (
+		<>
+			<div className="value-section__heading">{ __( 'Scan', 'jetpack-my-jetpack' ) }</div>
+			<div className="value-section__data">
+				<div>
+					<img
+						className="value-section__status-icon"
+						src={ ShieldOff }
+						alt={ __( 'Shield icon - Scan Status: Off', 'jetpack-my-jetpack' ) }
+					/>
+				</div>
+				<div className="value-section__status-text">{ __( 'Off', 'jetpack-my-jetpack' ) }</div>
+			</div>
+		</>
+	);
+};
```
There are many levels in this code. For readability, we could think of splitting it into e.g. small components to reduce complexity. Do you think it would be possible here?

elliottprogrammer replied:
Yes, I'll split into smaller components. 👍

robertsreberski commented:
For the testing plan: we can also modify plugin files in WP Admin directly (`/wp-admin/plugin-editor.php`)

I think we can remove that sentence from the component, I don't see it anywhere in designs:
![CleanShot 2024-07-09 at 14 52 53@2x](https://github.com/Automattic/jetpack/assets/8419292/231e1dc1-ebb1-4d7f-8fd8-b9cc085dbbbc)

For the free state card, the upgrade button is missing, but I'm not sure whether it should be part of this PR:
![CleanShot 2024-07-09 at 14 55 59@2x](https://github.com/Automattic/jetpack/assets/8419292/56465526-d065-421a-88d3-f2be3a7f8af5)

I couldn't get free version to show "Treats" information (following the testing plan) - I only see "Partial" information. Have I done something wrong or is it expected? If so, can we add point in the testing plan to be able to see one Treat in free version card?

Generally the PR tests well, and I see how complex it is to operate on these data. If it would help, we could move some of the logic that happens in the component to separate hook or `class-initializer.php` to make the view part simpler - what are your thoughts on this?

elliottprogrammer committed:
Remove unnecessary non-breaking space.

elliottprogrammer committed:
Remove unnecessary icon width & height properties is scss.

elliottprogrammer committed:
Combine singular/plural translations into one translated string, per feedback.

elliottprogrammer committed:
Split up scan-threats into smaller components for readability, per feedback.

elliottprogrammer committed:
Hide Protect card description, per feedback.

elliottprogrammer committed:
Show "Upgrade" button & secondary "View" button when no paid plan.

elliottprogrammer commented:
> For the testing plan: we can also modify plugin files in WP Admin directly (`/wp-admin/plugin-editor.php`)

Oh nice! I forgot about that!  Ok, thanks! 👍

> I think we can remove that sentence from the component (the product description), I don't see it anywhere in designs: !

Ok yeah, I was initially going to handle that in a future PR, but I went ahead and removed it in this PR. 👍

> For the free state card, the upgrade button is missing, but I'm not sure whether it should be part of this PR: 

Yeah, I was also going to tackle that in a future PR too, but I went ahead and handled it here is this PR too. 👍
> 
> I couldn't get free version to show "Treats" information (following the testing plan) - I only see "Partial" information. Have I done something wrong or is it expected? If so, can we add point in the testing plan to be able to see one Treat in free version card?

No that is not expected. I updated the testing instructions slightly to first only change the Akismet version (for the free scan), and then add the $eicar signature to trigger a critical threat (for the paid scan). I performed the updated steps myself and it worked as expected.

> Generally the PR tests well, and I see how complex it is to operate on these data. If it would help, we could move some of the logic that happens in the component to separate hook or `class-initializer.php` to make the view part simpler - what are your thoughts on this?

Yeah sure, I agree we could try to further simplify, but I think the effort should be applied **after** I get most all the overall functionality & logic of the entire card in place.  You see, I say this just because while I'm building this out, I'm often changing or even simplifying some things from the PR before it, and additionally, my main goal at first is to try to get a basic MVP done and in front of users as quickly as possible. Then after that, we can iterate and improve on it, over and over as much as we want. Thats was my thinking anyway. 🤷‍♂️ 😉

------------

Thanks so much for the review @robertsreberski!  I've addressed all your feedback. Would you mind taking a look again to confirm you don't see any other issues?  Much appreciated! 🙌

robertsreberski commented on projects/packages/my-jetpack/global.d.ts:
```diff
@@ -31,7 +31,7 @@ type ScanItem = {
 	name: string;
 	slug: string;
-	threats: string[];
+	threats: object[];
```
Can we define the object here, or the structure is unknown?

elliottprogrammer replied:
Yes, will do. 👍

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/style.scss:
```diff
@@ -57,4 +47,51 @@
+
+    &__status-text {
+        margin-right: 1px;
+        letter-spacing: -0.24px;
```
Do we need such specific spacing?

elliottprogrammer replied:
It was exactly per the design. I'll leave a comment. 👍

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/style.scss:
```diff
@@ -1,11 +1,14 @@
         font-size: var(--font-body-extra-small);
         color: var(--jp-gray-100);
         font-weight: 500;
+        margin-bottom: 10px;
```
can we reuse `--spacing-base`?

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/index.tsx:
```diff
@@ -1,20 +1,47 @@
+		shouldShowButton: shouldShowSecondaryButton,
 	};
 
+	const noDescription = useCallback( () => null, [] );
```
Maybe we can leave a comment for the future reference that if we gonna remove description to more cards, we should consider extending `<ProductCard />` functionality to support that.

Clever workaround here though! 🙌

robertsreberski commented on projects/packages/my-jetpack/src/products/class-protect.php:
```diff
@@ -255,16 +255,33 @@ public static function get_pricing_for_ui() {
+		if ( is_array( $purchases_data ) && ! empty( $purchases_data ) ) {
+			foreach ( $purchases_data as $purchase ) {
+				// Protect is available as jetpack_scan product and as part of the Security or Complete plan.
+				if ( strpos( $purchase->product_slug, 'jetpack_scan' ) !== false || str_starts_with( $purchase->product_slug, 'jetpack_security' ) || str_starts_with( $purchase->product_slug, 'jetpack_complete' ) ) {
```
We do we use `strpos()` for `jetpack_scan` product slug, but `str_starts_with()` for others?

Also, we could break the condition down to multiple lines, can be more readable that way 🤔

elliottprogrammer replied:
> We do we use `strpos()` for jetpack_scan product slug, but `str_starts_with()` for others?

I'm not quite sure, to be honest... This is just exactly how it is also done in many other of the product classes, so I just copied it over and did the same (just to play it safe). More specifically, the classes for Anti-spam, Boost, Creator, Search, Social, Stats, VideoPress, and now Protect all do it exactly like this in the `has_paid_plan_for_product()` function. 🤷‍♂️

Yes good idea, I'll break it down to be more readable. 👍

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/scan-threats.tsx:
```diff
@@ -0,0 +1,249 @@
+					/>
+				);
+			}
+			return <ScanStatus status={ 'success' } tooltipContent={ scanThreatsTooltip } />;
```
```suggestion
			return <ScanStatus status="success" tooltipContent={ scanThreatsTooltip } />;
```

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/scan-threats.tsx:
```diff
@@ -0,0 +1,249 @@
+		return numThreats ? (
+			<ThreatStatus numThreats={ numThreats } tooltipContent={ scanThreatsTooltip } />
+		) : (
+			<ScanStatus status={ 'partial' } tooltipContent={ scanThreatsTooltip } />
```
```suggestion
			<ScanStatus status="partial" tooltipContent={ scanThreatsTooltip } />
```

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/scan-threats.tsx:
```diff
@@ -0,0 +1,249 @@
+		);
+	}
+
+	return <ScanStatus status={ 'off' } tooltipContent={ scanThreatsTooltip } />;
```
```suggestion
	return <ScanStatus status="off" tooltipContent={ scanThreatsTooltip } />;
```

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/scan-threats.tsx:
```diff
@@ -0,0 +1,249 @@
+		}
+	}, [ setIsPopoverVisible, useTooltipRef ] );
+
+	return criticalThreatCount ? (
```
Can we still split it into two returns?

Like:

if ( criticalThreatCount ) {
   return <>Errors</>
}

return <>Default</>

Currently, it's easy to lose `:` operator in between these two layouts

robertsreberski reviewed (commented):
Code looks a lot better thanks!

Lovely that you collaborated with @CodeyGuyDylan on translation strings - the final idea is really neat!

I've left a bit more comments, but the code is almost there 🏁

elliottprogrammer committed:
Define Threat object in global.d.ts, per feedback.

elliottprogrammer committed:
Add scss comment explaining specific letter-spacing, per feedback

elliottprogrammer committed:
scss, use --spacing-base where possible.

elliottprogrammer committed:
Add comment about removing the <ProductCard /> Description, per feedback.

elliottprogrammer committed:
Split if condition into multiple lines, per feedback.

elliottprogrammer committed:
Remove ternary & split into 2 returns instead, per feedback.

elliottprogrammer commented:
> Also, thank you @elliottprogrammer for clarifying the instructions regarding testing. I do get 1 treat now on the free version. However, I do have a question. When I upgrade, and the initial scan is ongoing, should I see "Secure" state in the Protect card?
> 
> (See image above)
> 
> From what I understand, the previous treat (wrong version of Akismet) still persists on the website, why is it secure then?
> 
> It might be a question to the Scan team, but maybe you have an idea 🤔

@robertsreberski,
Yeah, I'm not quite sure why it does this. 
However, in an upcoming future PR when we incorporate a realtime loading/currently-scanning state and realtime data update, I believe the behavior you describe will no longer be apparent. 👍

elliottprogrammer commented:
> Code looks a lot better thanks!
> 
> Lovely that you collaborated with @CodeyGuyDylan on translation strings - the final idea is really neat!
> 
> I've left a bit more comments, but the code is almost there 🏁

Thanks again for the review @robertsreberski! I've addressed your additional feedback. Let me know if you see anything else or have any other suggestions for improvement. Thanks Robert!! 🙌

robertsreberski reviewed (dismissed):
Thanks for working on the updates @elliottprogrammer ! They look very good, and the PR tests well. 🟢

elliottprogrammer committed:
Merge branch 'trunk' into add/my-protect-card-scan-threats
